fix: NetworkSceneManager exception when DontDestroyOnLoad NetworkObjects are being synchronized

NoelStephensUnity opened the pull request:
[MTT-1136](https://jira.unity3d.com/browse/MTT-1136)
This fix pertains to the following[ GitHub-1088 issue](https://github.com/Unity-Technologies/com.unity.netcode.gameobjects/issues/1088).
This fix includes a validating manual test check for [Github-701 issue](https://github.com/Unity-Technologies/com.unity.multiplayer.mlapi/issues/701) in testproject\Assets\Tests\Manual\SceneTransitioning\SceneTransitioningTest that demonstrates how the NetworkSceneManager.OnSceneEvent can be used to detect when clients are ready to receive RPCs.

NoelStephensUnity committed:
fix

This fixes the dont destroy on load issue in NetworkSceneManager

NoelStephensUnity committed:
test

This provides manual tests to verify that don't destroy on load will work when NetworkManger's don't destroy property is enabled or disabled.

NoelStephensUnity committed:
fix

This should fix all scenarios with one scenario causing a warning message that might not be needed but I am leaving it in until I get a consensus on handling the scenario where a NetworkObject "will" be moved to the DontDestroyOnLoad scene but during the initial syncrhonization it does not exist.

0xFA11 committed:
Merge branch 'develop' into fix/develop-dontdestroy-networkscenemanager

NoelStephensUnity committed:
style and refactor

Changed some of the names used, added and updated comments, and applied the changes to the assets based on the name changes.

NoelStephensUnity committed:
Merge branch 'fix/develop-dontdestroy-networkscenemanager' of https://github.com/Unity-Technologies/com.unity.netcode.gameobjects into fix/develop-dontdestroy-networkscenemanager

VALERE91Unity commented on com.unity.netcode.gameobjects/Runtime/SceneManagement/NetworkSceneManager.cs:
```diff
@@ -245,13 +245,41 @@ internal void SetTheSceneBeingSynchronized(int serverSceneHandle)
                 // If the scene was not found (invalid) or was not loaded then throw an exception
```
Maybe removing this comment since we are not throwing anymore

NoelStephensUnity committed:
test

Adding the unit test to verify the baseline DontDestroyOnLoad scene synchronization is working when NetworkManager.DestroyOnLoad is enabled and disabled.

NoelStephensUnity committed:
style

updating comments based on PR suggestions.

mattwalsh-unity commented on com.unity.netcode.gameobjects/Runtime/SceneManagement/NetworkSceneManager.cs:
```diff
@@ -242,16 +242,44 @@ internal void SetTheSceneBeingSynchronized(int serverSceneHandle)
                 if (!SceneBeingSynchronized.IsValid() || !SceneBeingSynchronized.isLoaded)
                 {
-                    throw new Exception($"[{nameof(NetworkSceneManager)}- {nameof(ScenesLoaded)}] Could not find the appropriate scene to set as being synchronized!");
+                    // Let's go ahead and use the currently active scene under the scenario where a NetworkObject is determined to exist in a scene that the NetworkSceneManager is not aware of
```
So in the older code, we halted and caught fire if 1) the scene being sync'd was invalid or 2) it was not loaded.  

Case 1) Seems like we want to still halt and catch fire for situation 1).  Changing to some other scene would seem to mask a problem that needs to be dealt with.  I mean, why would we think we would want to pop back to the active scene if the scene we're going to is invalid.

Case 2) same kind of thing; if the scene is valid but not loaded yet, perhaps there is another handling - I'm just not sure why we'd just pick whatever scene happens to be loaded now

NoelStephensUnity replied:
This is specific to in-scene placed NetworkObjects during the client side deserialization phase for:

- Newly joined client synchronization where all scenes are already loaded and the client is just running through the in-scene placed NetworkObjects (deserialization) to be instantiated.  
- A level is loaded and the client is synchronizing in-scene placed NetworkObjects.

The only time the offending method (SetTheSceneBeingSynchronized) is truly pertinent is when you have the same additive scene loaded multiple times and there are in-scene NetworkObjects in each instance that need to be synchronized.  We use scene handlers to distinguish between "duplicate in-scene NetworkObjects" in order to "find the right NetworkObject" so we can apply the appropriate NetworkObjectId to that specific instance.
For all other cases, the GlobalObjectIdHash should be unique and will not matter really which scene is the current scene being synchronized.
What scenario would cause us to use the currently active scene?
If you set NetworkManager.DontDestroy to false and then have a NetworkObject that moves itself into the DontDestroyOnLoad scene when it is instantiated, then when a client is trying to "synchronize" to this specific scenario the first NetworkObject the client is trying to deserialize with this kind of behavior will not have a DontDestroyOnLoad scene (yet) and so it is perfectly fine to return the currently active scene.  Upon that first NetworkObject instantiating (after deserialization) the DontDestroyOnLoad scene will exist so any other NetworkObject being deserialized (with the same behavior) on the client from that point forward will set the DontDestroyOnLoad scene to be the "scene being synchronized".

It is one of those chicken or the egg scenarios:
DontDestroyOnLoad is only created when you move a GameObject into that scene using the GameObject.DontDestroyOnLoad method.  If nothing has been moved into the DontDestroyOnLoad scene, then it wont exist and isn't something that can be instantiated.  I had thought about adding a chunk of extra code that would tell the client to "pre-seed" the DontDestroyOnLoad scene (i.e. create a temporary GameObject before deserialization and then destroy it afterwards) but that seemed a bit over the top.  :)

mattwalsh-unity replied:
I think I may just need to trust you on this one, not sure I can parse this out / come to understand all the related systems as quickly as we need

NoelStephensUnity replied:
This was refactored a bit and might make more sense at this point.  But yeah.., there are a lot of different possible scenarios to consider for sure.

mattwalsh-unity commented on com.unity.netcode.gameobjects/Runtime/SceneManagement/NetworkSceneManager.cs:
```diff
@@ -242,16 +242,44 @@ internal void SetTheSceneBeingSynchronized(int serverSceneHandle)
+                }
+                else
+                {
+                    // The next scenario is if a user intentionally moves NetworkObjects into a DontDestroyOnLoad scene
```
When we say "user intentionally moves NO into a DDOL", is this like the user is manipulating objects manually in the editor while things are running?  I guess I  didn't think of this as something we needed to explicitly support.  Or is this  if *code* moves a NO in the DDOL scene?

NoelStephensUnity replied:
It is if the user's code intentionally moves it into the DDOL scene.
If you had a NetworkBehaviour that had this code in it:
```
        /// <summary>
        /// When enabled, we move ourself to the DontDestroyOnLoad scene
        /// </summary>
        private void OnEnable()
        {
            DontDestroyOnLoad(this);
        }
```
Then it would get moved into the DDOL Scene when it was instantiated... and that would apply to both dynamically spawned and in-scene placed NetworkObjects.

NoelStephensUnity committed:
Merge branch 'develop' into fix/develop-dontdestroy-networkscenemanager

NoelStephensUnity committed:
Merge branch 'develop' into fix/develop-dontdestroy-networkscenemanager

NoelStephensUnity committed:
fix

Fixing renamed method.

LukeStampfli commented on testproject/Assets/Tests/Manual/DontDestroyOnLoad/DontDestroyOnLoadSpawnHandler.cs:
```diff
@@ -0,0 +1,25 @@
+    /// <summary>
+    /// Spawns a single NetworkPrefab
+    /// </summary>
+    public class DontDestroyOnLoadSpawnHandler : NetworkBehaviour
```
nit: This really has nothing todo with DontDestroyOnLoad so maybe we could just call this `SingleNetworkObjectSpawner` or something like that?

NoelStephensUnity replied:
Yeah, I had that thought but then came to the conclusion since the script is contained within the Tests/Manual/DontDestroyOnLoad folder that it was "ok" to name it that since its primary function is to do nothing more than spawn the object that places itself in the DontDestroyOnLoad scene.

LukeStampfli commented on testproject/Assets/Tests/Manual/DontDestroyOnLoad/ObjectToNotDestroyBehaviour.cs:
```diff
@@ -0,0 +1,93 @@
+        /// </summary>
+        private void OnEnable()
+        {
+            DontDestroyOnLoad(this);
```
So it looks like we have 2 features now which do something similar.
- First of all a user can call DontDestroyOnLoad on a NetworkObject to move it into the `DontDestroyOnLoad` scene and we support this.
- NetworkObject.Spawn takes a `destroyWithScene` parameter which can be set to false. This acts where similar to `DontDestroyOnLoad` but keeps the object in the original scene and does not move it into the `DontDestroyOnLoad` scene.

Now putting my user hat on. Which one of the two should I use in which case? Why are there two ways of doing this? Do we think we need education or documentation to solve this? Or do we think we should only have one feature for solving this?

LukeStampfli replied:
[Spawn function for reference](https://github.com/Unity-Technologies/com.unity.netcode.gameobjects/blob/ee59d5fb560bbd6273260c1df144172914b38448/com.unity.multiplayer.mlapi/Runtime/Core/NetworkObject.cs#L401)

NoelStephensUnity replied:
The DontDestroyOnLoad method was already there...I am just using it.  :)
However, here are the scenarios:
I have in-scene placed NetworkObject that I want to move into the DontDestroyOnLoad scene when it is instantiated.  I can't spawn it because the level is loading it and it is being "spawned" that way... so I call DontDestroyOnLoad(this) which is the base functionality we inherited from [Object.DontDestroyOnLoad](https://docs.unity3d.com/ScriptReference/Object.DontDestroyOnLoad.html)
Then we have the scenario where we want to spawn something and let clients know they should place it in the DontDestroyOnLoad scene "automatically"... thus we have the [NetworkObject.Spawn](https://github.com/Unity-Technologies/com.unity.netcode.gameobjects/blob/ee59d5fb560bbd6273260c1df144172914b38448/com.unity.multiplayer.mlapi/Runtime/Core/NetworkObject.cs#L401) method.

So, there are two ways a user can do it:
1. Use the method that is built into the Unity Object class
2. Use the NetworkObject.Spawn method that is built into the Netcode for GameObjects class

I think both are valid and a user should be able to use both paths (really we can't prevent them from using #1) and it should "just work".   Does that make sense?

LukeStampfli replied:
Thanks for the explanation. I think what you said makes complete sense and I fully agree 👍

LukeStampfli commented on com.unity.netcode.gameobjects/Runtime/SceneManagement/NetworkSceneManager.cs:
```diff
@@ -242,16 +242,44 @@ internal void SetTheSceneBeingSynchronized(int serverSceneHandle)
             {
-                // This should never happen, but in the event it does...
-                throw new Exception($"[{nameof(SceneEventData)}- Scene Handle Mismatch] {nameof(serverSceneHandle)} could not be found in {nameof(ServerSceneHandleToClientSceneHandle)}!");
+                // Most common scenario for DontDestroyOnLoad is when NetworkManager is set to not be destroyed
```
Not sure whether I understand this. From my understanding `SetTheSceneBeingSynchronized` is being called when a scene NetworkObject is being serialized. Now that object could be in any scene including the DontDestroyOnLoad scene. But why is there this special case here. From my understanding this assumes that if the `NetworkManager` is `DontDestroyOnLoad` all invalid server scene handles are automatically translated to the `DontDestroyOnLoad` scene. Isn't that wrong?

NoelStephensUnity replied:
Really you have to look at the full context:
```
             // For the client, we make a server scene handle to client scene handle look up table
                if (!ServerSceneHandleToClientSceneHandle.ContainsKey(SceneEventData.SceneHandle))
                {
                    ServerSceneHandleToClientSceneHandle.Add(SceneEventData.SceneHandle, nextScene.handle);
                }
                else
                {
                    // If the exact same handle exists then there are problems with using handles
                    throw new Exception($"Server Scene Handle ({SceneEventData.SceneHandle}) already exist!  Happened during scene load of {nextScene.name} with Client Handle ({nextScene.handle})");
                }
```
What we are saying is this:

- Check to make sure you haven't already registered this server scene handle for the scene you just loaded
- If you have the server scene handle already, then something very bad happened and it is a bug
  - Meaning, this exception should never happen unless there is some really odd-ball combination of events that occur which I did not take into account...because at that point in the code we are a client and we have just finished loading a scene and want to associate our local Scene.handle with the server's Scene.Handle for the scene just loaded...which we will use later (this all has to do with being able to load the same additive scene multiple times).

LukeStampfli commented on com.unity.netcode.gameobjects/Runtime/SceneManagement/NetworkSceneManager.cs:
```diff
@@ -242,16 +242,44 @@ internal void SetTheSceneBeingSynchronized(int serverSceneHandle)
+
+                    // Let's go ahead and use the currently active scene under the scenario where a NetworkObject is determined to exist in a scene that the NetworkSceneManager is not aware of
+                    // or the NetworkObject has yet to be moved to that specific scene (i.e. no DontDestroyOnLoad scene exists yet).
+                    SceneBeingSynchronized = SceneManager.GetActiveScene();
```
Won't that cause long term issues and confusion? E.g. the object will permanently end up in the wrong scene. Could we savely throw an exception here?

NoelStephensUnity replied:
Yeah, this spot has raised a few eyebrows but actually it makes sense...but we need to look at the full context again:
```
                   // The next scenario is if a user intentionally moves NetworkObjects into a DontDestroyOnLoad scene
                    var dontDestroyOnLoadScene = SceneManager.GetSceneByName("DontDestroyOnLoad");
                    if (dontDestroyOnLoadScene.IsValid() && dontDestroyOnLoadScene.isLoaded && dontDestroyOnLoadScene.handle == serverSceneHandle)
                    {
                        SceneBeingSynchronized = dontDestroyOnLoadScene;
                        return;
                    }

                    // Let's go ahead and use the currently active scene under the scenario where a NetworkObject is determined to exist in a scene that the NetworkSceneManager is not aware of
                    // or the NetworkObject has yet to be moved to that specific scene (i.e. no DontDestroyOnLoad scene exists yet).
                    SceneBeingSynchronized = SceneManager.GetActiveScene();
```
This scenario will crop up on the client side if you set NetworkManager.DontDestroy to false.  The primary reason is that the DontDestroyOnLoad scene is not created until you move something into it.  Let me provide a quick example:

1. You have an in-scene NetworkObject that moves itself into the DontDestroyOnLoad scene when instantiated and you have disabled NetworkManager.DontDestroyOnLoad.
2. You start a host or server and the in-scene NetworkObject instantiates and moves itself into the DontDestroyOnLoad scene.
3. You connect a client that runs through the synchronization process and loads all scenes except for the DontDestroyOnLoad because that can only be "loaded" when you move something into it.
4. Upon finishing the scene loading, the client then begins to synchronize in-scene placed NetworkObjects (deserialization).
5. When the client runs across a NetworkObject who's associated scene  can't be found in its look up table it then tries to first find the "DontDestroyOnLoad" scene which is not loaded/created yet.  The SceneBeingSynchronized is only really "pertinent" when you have multiple instances of the same additive scene and you are trying to determine "which of the (n) in-scene placed NetworkObjects with the exact same name and GlobalObjectIdHash values are we trying to synchronize?"
6. The way we are handling that is associating scene handles with NetworkObjects (i.e. the scene the GameObject belongs to is the associated scene's handle).  
7. Regarding the DontDestroyOnLoad scene, we are making the assumption/rule that you would never load an additive scene multiple times that will produce the exact same named in-scene placed NetworkObjects with the exact same GlobalObjectIdHash value and then move each instance of the identical additive scene relative in-scene placed NetworkObjects into the DontDestroyOnLoad scene.
8. So...you have a client that now is being told that a NetworkObject belongs to the DontDestroyOnLoad scene (this is only for the very first NetworkObject that belongs to the DontDestroyOnLoad scene) and since it isn't a scene the NetworkSceneManager is aware of (yet) it assigns the default currently actives scene as the SceneBeingSynchronized.  Since any other "deserialization" of NetworkObjects that aren't created in an additive scene that is loaded multiple times (i.e. NetworkSceneManager would be aware of this scene handle) doesn't depend upon the scene handle to find the unique GlobalObjectIdHash value for the NetworkObject we are going to be deserializing... we can use the currently active scene.
9. When the NetworkObject is deserialized and instantiated it will move itself into the DontDestroyOnLoad scene which, in turn, will create the DontDestroyOnLoad scene...and so any other NetworkObject that might move into the DontDestroyOnLoad scene will not use the currently active scene because they will then, at that point, be able to find the DontDestroyOnLoad scene.

This is only under the scenario where a user has opted to set the NetworkManager.DontDestroy to false because when it is true the NetworkManager itself will have already caused the DontDestroyOnLoad scene to be created (by the time you get to this point in the code it will already exist). 

So, it won't cause any issues down the road...it is just a really "rare case" scenario that I am taking into account so we are able to fully support DontDestroyOnLoad usage patterns when the NetworkManager.DontDestroy is set to false.

(phew) 
It is a rather lengthy explanation...but then again it is a semi-complex thing to explain too.  :)

LukeStampfli replied:
Thanks for explaining, what you are saying makes sense and it's indeed quite a complicated topic. I did not think about the scenario where the DontDestroyOnLoad scene was not created yet.

NoelStephensUnity committed:
fix

Verifying alternate DontDestroyOnLoad test case mixed with checking when we can or cannot send RPCs to a client.

NoelStephensUnity committed:
refactor

Cleaning up our DDOL GameObject.

NoelStephensUnity committed:
Merge branch 'develop' into fix/develop-dontdestroy-networkscenemanager

NoelStephensUnity committed:
refactor and style

Removed the NetworkObject property and assignment code from the PlayerMovementManager as it is not needed.

Updated comments in NetworkSceneManager regarding DDOL.

JesseOlmer commented on com.unity.netcode.gameobjects/Runtime/SceneManagement/NetworkSceneManager.cs:
```diff
@@ -187,6 +192,47 @@ internal NetworkSceneManager(NetworkManager networkManager)
+            else // Otherwise, we have to create a GameObject and move it into the DDOL to get the scene
+            {
+
+#if UNITY_EDITOR || DEVELOPMENT_BUILD
```
Use UNITY_INCLUDE_TESTS instead of these...

NoelStephensUnity replied:
@JesseOlmer I am going to fix those in PR-1080 if that is ok?

andrews-unity commented on com.unity.netcode.gameobjects/Runtime/SceneManagement/NetworkSceneManager.cs:
```diff
@@ -187,6 +192,47 @@ internal NetworkSceneManager(NetworkManager networkManager)
+        /// <summary>
+        /// Handle cleaning up the DDOL GameObject
+        /// </summary>
+        public void Dispose()
```
Is this called anywhere I could not see it being called?

NoelStephensUnity replied:
When it is destroyed? 
` public class NetworkSceneManager:IDisposable`

NoelStephensUnity replied:
This is removed with the most recent commit.

andrews-unity commented on com.unity.netcode.gameobjects/Runtime/SceneManagement/NetworkSceneManager.cs:
```diff
@@ -187,6 +192,47 @@ internal NetworkSceneManager(NetworkManager networkManager)
+#endif
+                // Create our DDOL GameObject and move it into the DDOL scene so we can register the DDOL with
+                // the NetworkSceneManager.
+                m_DDOLObject = new GameObject("DDOL-NWSM");
```
can we just cache off the scene handle and then destroy the object right away? thus not polluting the scene with a useless object ?

NoelStephensUnity replied:
Let me see if that works.

NoelStephensUnity replied:
It indeed did work to do it that way!  ⚡ ZAPPED that code out of there!

NoelStephensUnity committed:
refactor

We don't need to keep the DDOL object around after we have the DDOL scene.
Removed IDisposable from NetworkSceneManager
Removed the private m_DDOLObject
Removed the Dispose method.

NoelStephensUnity committed:
Merge branch 'develop' into fix/develop-dontdestroy-networkscenemanager
